### PR TITLE
use an object with functions as "RPC message" type

### PR DESCRIPTION
**Pull request by phryneas**
Triggered by https://github.com/apollographql/apollo-client-devtools/pull/1256#pullrequestreview-1926990431

### CODE REVIEW

File: src/extension/rpc.ts [hunk 1]
```diff
@@ -17,46 +17,34 @@ type RPCResponseMessageFormat =
...
-export type RPCMessage<
-  Name extends string,
-  Params extends RPCParams,
-  ReturnType,
-> = {
-  __name: Name;
-  __params: Params;
-  __returnType: ReturnType;
-};
+type MessageCollection = Record<string, (...parameters: [SafeAny]) => SafeAny>;
```
Comment (phryneas) [inline src/extension/rpc.ts:20]: Using `RPCParams` here unfortunately causes all kind of downstream problems, especially with the "empty" case of `Record<string, never>` - so I've gone for a `SafeAny` here. Is there a specific reason this has to be an object and not just any serializable value>

Comment (jerelmiller) [reply src/extension/rpc.ts:20]: No reason it had to be an object. This was one of the final improvements I hadn't come back to :). I like this better.

File: src/extension/rpc.ts [hunk 2]
```diff
@@ -17,46 +17,34 @@ type RPCResponseMessageFormat =
...
 
-export interface RpcClient<
-  Messages extends RPCMessage<string, RPCParams, SafeAny>,
-> {
-  request: <TName extends Messages["__name"]>(
+export interface RpcClient<Messages extends MessageCollection> {
+  request: <TName extends keyof Messages & string>(
```
Comment (jerelmiller) [inline src/extension/rpc.ts:23]: I'm curious the reason behind the `& string` here. Is that just to filter out any potential keys that aren't strings?

**Review by jerelmiller — APPROVED** [review]
I like this a lot. This feels so much more natural than the way I had set it up. Thanks so much for the improvement here! I'll keep iterating a bit to see if we can figure out how to make `params` a bit more natural, but otherwise this is amazing.

File: src/extension/rpc.ts (hunk 2 above)
Comment (phryneas) [reply src/extension/rpc.ts:23]: Yeah, otherwise `keyof` will always be `| symbol`, which makes using this in other places very annoying.

**Pull request merged**